File: src/main/java/iot/tetracube/deviceComunication/deviceProvisioning/DeviceProvisioningService.java
package iot.tetracube.deviceComunication.deviceProvisioning;

import io.quarkus.vertx.ConsumeEvent;
import io.smallrye.mutiny.Multi;
import io.smallrye.mutiny.Uni;
import io.vertx.core.logging.Logger;
import io.vertx.core.logging.LoggerFactory;
import iot.tetracube.data.entities.Action;
import iot.tetracube.data.entities.Device;
import iot.tetracube.data.repositories.ActionRepository;
import iot.tetracube.data.repositories.DeviceRepository;
import iot.tetracube.models.dto.ManageDeviceProvisioningResponse;
import iot.tetracube.deviceComunication.deviceProvisioning.payloads.DeviceActionProvisioningMessage;
import iot.tetracube.deviceComunication.deviceProvisioning.payloads.DeviceProvisioningMessage;

import javax.enterprise.context.ApplicationScoped;
import javax.json.bind.Jsonb;
import java.util.List;
import java.util.Objects;
import java.util.UUID;

@ApplicationScoped
public class DeviceProvisioningService {

    private final static Logger LOGGER = LoggerFactory.getLogger(DeviceProvisioningService.class);

    private final Jsonb jsonb;
    private final DeviceRepository deviceRepository;
    private final ActionRepository actionRepository;

    public DeviceProvisioningService(Jsonb jsonb,
                                     DeviceRepository deviceRepository,
                                     ActionRepository actionRepository) {
        this.jsonb = jsonb;
        this.deviceRepository = deviceRepository;
        this.actionRepository = actionRepository;
    }

    @ConsumeEvent("device-provisioning")
    public Uni<ManageDeviceProvisioningResponse> manageDeviceProvisioningMessage(String message) {
        LOGGER.info("Parsing device provisioning message");
        DeviceProvisioningMessage deviceProvisioningMessage;
        try {
            deviceProvisioningMessage = this.jsonb.fromJson(message, DeviceProvisioningMessage.class);
        } catch (Exception e) {
            LOGGER.error("Cannot convert message, ignoring it");
            return Uni.createFrom().nullItem();
        }

        LOGGER.info("Checking if device already exists");
        var deviceExistsUni = this.deviceRepository.deviceExistsByCircuitId(deviceProvisioningMessage.getId());

        LOGGER.info("Save or update existing device");
        var deviceUni = deviceExistsUni.flatMap(deviceExists -> {
            if (deviceExists == null) {
                LOGGER.error("There was some error during device verification, returning bad feedback to device");
                return Uni.createFrom().nullItem();
            } else if (deviceExists) {
                LOGGER.info("Device exists, updating only its actions if needed");
                return this.deviceRepository.getDeviceByCircuitId(deviceProvisioningMessage.getId());
            } else {
                LOGGER.info("Storing device data in database");
                var deviceEntity = new Device(
                        UUID.randomUUID(),
                        deviceProvisioningMessage.getId(),
                        deviceProvisioningMessage.getDefaultName(),
                        false,
                        deviceProvisioningMessage.getHostname()
                );
                return this.deviceRepository.saveDevice(deviceEntity);
            }
        });

        LOGGER.info("Store device actions");
        var actionSavedResultUni = deviceUni.flatMap(device -> {
            if (device == null) {
                return Uni.createFrom().item(new ManageDeviceProvisioningResponse(deviceProvisioningMessage.getId(), false));
            }
            LOGGER.info("Device stored, now processing actions");
            return this.storeDeviceAction(device, deviceProvisioningMessage.getDeviceActionProvisioningMessages())
                    .map(result -> new ManageDeviceProvisioningResponse(device.getCircuitId(), result));
        });

        LOGGER.info("Subscribing to events");
        actionSavedResultUni
                .invoke(device -> {
                    LOGGER.info("Processing promises chain");
                });

        return actionSavedResultUni;
    }

    private Uni<Boolean> storeDeviceAction(Device parentDevice, List<DeviceActionProvisioningMessage> actionsToStore) {
        LOGGER.info("Processing " + actionsToStore.size() + " actions of device id " + parentDevice.getId());
        return Multi.createFrom().items(actionsToStore.stream())
                .onItem()
                .transformToMulti(actionToStore ->
                        this.manageActionProvisioningMessage(parentDevice.getId(), actionToStore).toMulti()
                )
                .concatenate().collectItems().asList()
                .map(responses -> responses.stream().allMatch(response -> response));
    }

    private Uni<Boolean> manageActionProvisioningMessage(UUID parentDeviceId, DeviceActionProvisioningMessage actionToStore) {
        var actionExistsUni = this.actionRepository.existsActionByHardwareId(actionToStore.getId());
        return actionExistsUni.flatMap(actionExists -> {
            if (actionExists == null) {
                return Uni.createFrom().item(false);
            } else if (actionExists) {
                LOGGER.info("Ignoring action: already exists");
                return Uni.createFrom().item(true);
            } else {
                LOGGER.info("Adding action");
                var action = new Action(
                        UUID.randomUUID(),
                        parentDeviceId,
                        actionToStore.getTopic(),
                        actionToStore.getDefaultName(),
                        actionToStore.getId()
                );
                return this.actionRepository.saveAction(action)
                        .map(Objects::nonNull);
            }
        });
    }

}

File: src/main/kotlin/iot/tetracubered/services/DeviceProvisioningService.kt
package iot.tetracubered.services

import io.quarkus.vertx.ConsumeEvent
import io.smallrye.mutiny.Multi
import io.smallrye.mutiny.Uni
import io.vertx.mutiny.core.eventbus.EventBus
import iot.tetracubered.data.entities.Action
import iot.tetracubered.data.entities.Device
import iot.tetracubered.data.repositories.ActionRepository
import iot.tetracubered.data.repositories.DeviceRepository
import iot.tetracubered.messaging.DeviceMessagingHub
import iot.tetracubered.messaging.payloads.ActionProvisioningPayload
import iot.tetracubered.messaging.payloads.DeviceFeedbackMessage
import iot.tetracubered.messaging.payloads.DeviceProvisioningPayload
import org.slf4j.LoggerFactory
import java.util.*
import javax.enterprise.context.ApplicationScoped

@ApplicationScoped
class DeviceProvisioningService(
    private val deviceRepository: DeviceRepository,
    private val actionRepository: ActionRepository,
    private val eventBus: EventBus
) {

    private val logger = LoggerFactory.getLogger(DeviceProvisioningService::class.java)

    @ConsumeEvent("device-provisioning")
    fun handleDeviceProvisioning(deviceProvisioningPayload: DeviceProvisioningPayload) {
        logger.info("Checking if device id ${deviceProvisioningPayload.circuitId} exists")
        this.deviceRepository.deviceExistsByCircuitId(circuitId = deviceProvisioningPayload.circuitId)
            .flatMap { existsResponse -> this.getDeviceOrCreate(deviceProvisioningPayload, existsResponse) }
            .flatMap { device -> this.manageActionsProvisioning(device, deviceProvisioningPayload.actions) }
            .subscribe()
            .with {
                logger.info("Device provisioning subscribed")
                this.eventBus.publish("send-feedback",
                    DeviceFeedbackMessage(
                        deviceProvisioningPayload.circuitId,
                        deviceProvisioningPayload.feedbackTopic,
                        true
                    )
                )
            }
    }

    private fun getDeviceOrCreate(deviceProvisioningPayload: DeviceProvisioningPayload, exists: Boolean): Uni<Device> =
        if (exists) {
            this.deviceRepository.getDeviceByCircuitId(deviceProvisioningPayload.circuitId)
        } else {
            this.deviceRepository.createDevice(
                Device(
                    UUID.randomUUID(),
                    deviceProvisioningPayload.circuitId,
                    deviceProvisioningPayload.name,
                    false,
                    deviceProvisioningPayload.feedbackTopic,
                    deviceProvisioningPayload.queryingTopic,
                    "",
                    null,
                    emptyList()
                )
            )
        }

    private fun manageActionsProvisioning(device: Device, actions: List<ActionProvisioningPayload>) =
        Multi.createFrom().items(actions.stream())
            .flatMap { actionProvisioning ->
                this.actionRepository.actionExistsByActionId(device.circuitId, actionProvisioning.actionId)
                    .flatMap { exists -> this.createOrGetAction(device, actionProvisioning, exists) }
                    .toMulti()
            }
            .collectItems().asList()

    private fun createOrGetAction(device: Device, action: ActionProvisioningPayload, exists: Boolean): Uni<Action> =
        if (exists) {
            this.actionRepository.getActionByActionId(device.id, action.actionId)
        } else {
            this.actionRepository.saveAction(
                Action(
                    UUID.randomUUID(),
                    action.actionId,
                    action.name,
                    action.commandTopic,
                    null,
                    device.id
                )
            )
        }
}